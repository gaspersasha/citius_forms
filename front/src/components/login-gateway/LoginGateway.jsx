import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { AuthModal } from '~containers/authentication';
import { Button, Typography, Spinner } from '~components';
import { useUserContext } from '~contexts';
import s from './login-gateway.module.sass';

const LoginGateway = ({ children }) => {
  const { user } = useUserContext();
  const [isModalOpen, setModalOpen] = useState(false);
  const [formType, setFormType] = useState('login');

  const handleClick = () => {
    setFormType('login');
    setModalOpen(true);
  };

  const render = user.isLoggedIn ? (
    children
  ) : (
    <>
      {isModalOpen && (
        <AuthModal
          onModalClose={() => setModalOpen(false)}
          activeForm={formType}
          setActiveForm={(form) => setFormType(form)}
        />
      )}
      <div className={s.wrapper}>
        <Typography type="h2">Please log in</Typography>
        <Typography className={s.msg}>
          You need to be logged in to access this page, please log in via the
          link below
        </Typography>
        <Button onClick={handleClick}>Log in</Button>
      </div>
    </>
  );

  return user.isLoading ? <Spinner /> : render;
};

LoginGateway.propTypes = {
  children: PropTypes.node.isRequired,
};

export default LoginGateway;
