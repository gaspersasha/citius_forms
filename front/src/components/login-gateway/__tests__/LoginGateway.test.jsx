import React from 'react';
import LoginGateway from '../LoginGateway';

describe('<LoginGateway />', () => {
  let wrapper;

  const getWrapper = (isLoggedIn = true) => {
    global.mockUseUserContext({
      user: { isLoggedIn },
    });

    return shallow(
      <LoginGateway>
        <div>Test</div>
      </LoginGateway>
    );
  };

  it('renders children in logged in user case', () => {
    wrapper = getWrapper();
    expect(wrapper).toMatchSnapshot();
  });

  it('renders without errors', () => {
    wrapper = getWrapper(false);
    expect(wrapper).toMatchSnapshot();
  });

  it('handles login modal opening case', () => {
    wrapper = getWrapper(false);
    wrapper.find('Button').simulate('click');

    const modal = wrapper.find('AuthModal');

    expect(modal.prop('activeForm')).toEqual('login');
    expect(modal.exists()).toEqual(true);
  });

  it('handles login modal closing case', () => {
    wrapper = getWrapper(false);
    wrapper.find('Button').simulate('click');

    wrapper.find('AuthModal').invoke('onModalClose')();
    expect(wrapper.find('AuthModal').exists()).toBe(false);
  });

  it('handles modal form type change', () => {
    wrapper = getWrapper(false);
    wrapper.find('Button').simulate('click');

    wrapper.find('AuthModal').invoke('setActiveForm')('reset');
    expect(wrapper.find('AuthModal').prop('activeForm')).toEqual('reset');
  });
});
