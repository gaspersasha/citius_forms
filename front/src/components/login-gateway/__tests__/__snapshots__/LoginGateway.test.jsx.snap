// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginGateway /> renders children in logged in user case 1`] = `
<div>
  Test
</div>
`;

exports[`<LoginGateway /> renders without errors 1`] = `
<Fragment>
  <div
    className="wrapper"
  >
    <Typography
      align="left"
      className=""
      color="dark"
      type="h2"
    >
      Please log in
    </Typography>
    <Typography
      align="left"
      className="msg"
      color="dark"
      type="p"
    >
      You need to be logged in to access this page, please log in via the link below
    </Typography>
    <Button
      className=""
      disabled={false}
      href=""
      onClick={[Function]}
      selected={false}
      styleType="primary"
    >
      Log in
    </Button>
  </div>
</Fragment>
`;
